docs: update contributing guidelines (#10223)
cosmos/cosmos-sdk: CONTRIBUTING.md

        - [Requesting Reviews](#requesting-reviews)
        - [Reviewing Pull Requests](#reviewing-pull-requests)
        - [Updating Documentation](#updating-documentation)
    - [Dependencies](#dependencies)
    - [Protobuf](#protobuf)
    - [Testing](#testing)
    - [Branching Model and Release](#branching-model-and-release)
        - [PR Targeting](#pr-targeting)
        - [Development Procedure](#development-procedure)
        - [Major Release Procedure](#major-release-procedure)
        - [Minor Release Procedure](#minor-release-procedure)
    - [Code Owner Membership](#code-owner-membership)
    - [Concept & Feature Approval Process](#concept--feature-approval-process)

> concept--release-approval-process

Reviewer: we can omit this line prob?

Author: sure. removing it.

        - [Requesting Reviews](#requesting-reviews)
        - [Reviewing Pull Requests](#reviewing-pull-requests)
        - [Updating Documentation](#updating-documentation)
    - [Dependencies](#dependencies)
    - [Protobuf](#protobuf)
    - [Testing](#testing)
    - [Branching Model and Release](#branching-model-and-release)
        - [PR Targeting](#pr-targeting)
        - [Development Procedure](#development-procedure)
        - [Major Release Procedure](#major-release-procedure)
        - [Minor Release Procedure](#minor-release-procedure)
    - [Code Owner Membership](#code-owner-membership)
    - [Concept & Feature Approval Process](#concept--feature-approval-process)

Thank you for considering making contributions to Cosmos SDK and related
repositories!

Contributing to this repo can mean many things such as participating in
discussion or proposing code changes. To ensure a smooth workflow for all
contributors, the general procedure for contributing has been established:

> 1. Either [open](https://github.com/cosmos/cosmos-sdk/issues/new/choose) or

Author: Adding 
_1. Start by browsing [new issue](https://github.com/cosmos/cosmos-sdk/issues) and [discussions](https://github.com/cosmos/cosmos-sdk/discussions). If you are looking for something interesting or if you have something in your mind, there is a chance it was has been discussed._
To the list.

        - [Requesting Reviews](#requesting-reviews)
        - [Reviewing Pull Requests](#reviewing-pull-requests)
        - [Updating Documentation](#updating-documentation)
    - [Dependencies](#dependencies)
    - [Protobuf](#protobuf)
    - [Testing](#testing)
    - [Branching Model and Release](#branching-model-and-release)
        - [PR Targeting](#pr-targeting)
        - [Development Procedure](#development-procedure)
        - [Major Release Procedure](#major-release-procedure)
        - [Minor Release Procedure](#minor-release-procedure)
    - [Code Owner Membership](#code-owner-membership)
    - [Concept & Feature Approval Process](#concept--feature-approval-process)

> concept--release-approval-process
> 

Author: ```suggestion
```

contributors, the general procedure for contributing has been established:

1. Either [open](https://github.com/cosmos/cosmos-sdk/issues/new/choose) or

> 1. Weather:

Reviewer: ```suggestion
1. Whether:
```

   1. you want to propose something new, which requires specification, additional design or you would like to change a process, start with a [new discussion](https://github.com/cosmos/cosmos-sdk/discussions/new). With discussions we can better handle the design process using discussion threads. A discussion should usually lead to one or few issues,
   2. the issue is a specified proposal or a bug then open directly a [new issue](https://github.com/cosmos/cosmos-sdk/issues/new/choose),

>    3. [find](https://github.com/cosmos/cosmos-sdk/issues) an issue you'd like to help with.

Author: Good suggestion!

  - For core developers working within the `cosmos-sdk` repo, follow branch name conventions to ensure a clear
    ownership of branches: `{moniker}/{issue#}-branch-name`.
  - See [Branching Model](#branching-model-and-release) for more details.
- Please make sure to run `make format` before every commit - the easiest way
  to do this is have your editor run it for you upon saving a file (most of the editors
  will do it anyway using a pre-configured setup of the programming language mode).

>   Additionally please ensure that your code is lint compliant by running `make lint-fix`.

Reviewer 2: ```suggestion
- Be sure to run `make format` before every commit. The easiest way
  to do this is have your editor run it for you upon saving a file (most of the editors
  will do it anyway using a pre-configured setup of the programming language mode).
  Additionally, be sure that your code is lint compliant by running `make lint-fix`.
```

### Requesting Reviews

In order to accommodate the review process, the author of the PR must complete the author checklist (from the pull request template)

> to the best of their abilities before marking the PR as "Ready for Review". If you would like to

Author: at the beginning we have "TL;RD". Here we go more in details. Do you think we should remove the section at the top?

> to the best of their abilities before marking the PR as "Ready for Review". If you would like to

Reviewer 2: seems redundant to me, consider linking from the TL;DR from this section. 

(I always thought that TL;DR was acronym for "too long didn't read"... teach me otherwise)

> to the best of their abilities before marking the PR as "Ready for Review". If you would like to

Author: OK, I moved and merged the PR content from "TL;DR" to here.

<some table>
for tcIndex, tc := range cases {
  <some code>
  resp, err := doSomething()
  require.NoError(err)
  require.Equal(t, tc.expected, resp, "should correctly perform X")
```

## Branching Model and Release

User-facing repos should adhere to the trunk based development branching model: https://trunkbaseddevelopment.com/. User branches should start with a user name, example: `{moniker}/{issue#}-branch-name`.

The Cosmos SDK repository is a [multi Go module](https://github.com/golang/go/wiki/Modules#is-it-possible-to-add-a-module-to-a-multi-module-repository) repository. It means that we have more than one Go module in a single repository.

The Cosmos SDK utilizes [semantic versioning](https://semver.org/).

### PR Targeting

Ensure that you base and target your PR on the `master` branch.

> All feature additions must be targeted against `master`. Bug fixes for an outstanding release candidate

Author: Yes. Only bug-fixes which are related only to the release candidate must target the release branch

> All feature additions must be targeted against `master`. Bug fixes for an outstanding release candidate

Author: Updated to provide more details about targeting PRs against release branch.

must be targeted against the release candidate branch.

If needed, we backport a commit from `master` to the release branch (excluding consensus breaking feature, optimization, and similar).

### Development Procedure

- the latest state of development is on `master`
- `master` must never fail `make lint test test-race`
- no `--force` onto `master` (except when reverting a broken commit, which should seldom happen)
- create a development branch either on github.com/cosmos/cosmos-sdk, or your fork (using `git remote add origin`)

Submitting a pull request:
- before submitting a pull request, merge the latest master `git merge origin/master`,
- run `make lint test` to ensure that all checks and tests pass,
- GitHub will squash commits and rebase on top of the master.

Finalizing pull requests
- ask contributors for review,
- respond to all comments,
- resolve merge conflicts

> Pull requests are merged automatically using `automerge` action.

Author: added a link to https://mergify.io/features/auto-merge

- In the release branch, prepare a new version section in the `CHANGELOG.md`
    - All links must be link-ified: `$ python ./scripts/linkify_changelog.py CHANGELOG.md`
    - Copy the entries into a `RELEASE_CHANGELOG.md`, this is needed so the bot knows which entries to add to the release page on GitHub.
- Create a new annotated git tag for a release candidate  (eg: `git tag -a v1.1.0-rc1`) in the release branch.
  - from this point we unfreeze master.
  - the SDK teams collaborate and do their best to run testnets in order to validate the release.
  - when bugs are found, create a PR for `master`, and backport fixes to the release branch.
  - create new release candidate tags after bugs are fixed.
- After the team feels the release branch is stable and everything works, create a full release:
  - update `CHANGELOG.md`.
  - create a new annotated git tag (eg `git -a v1.1.0`) in the release branch.
  - Create a GitHub release.

### Minor Release Procedure

A _minor release_ is an increment of the patch number (eg: `v1.2.0` → `v1.2.1`).

Following _semver_ philosophy, point releases after `v1.0`:

 - must not break API
 - can break consensus

> Patch release must not break API nor consensus.

Author: It's noted at the beginning of the paragraph:
> A _minor release_ is an increment of the patch number ...

I will use a _Patch_ terminology.

Before `v1.0`, point release can break both point API and consensus.

Updates to the release branch should come from `master` by backporting PRs (usually done by automatic cherry pick followed by a PRs to the release branch). The backports must be marked using `backport/Y` label in PR for master.
It is the PR author's responsibility to fix merge conflicts, update changelog entries, and
ensure CI passes. If a PR originates from an external contributor, a core team member assumes 
responsibility to perform this process instead of the original author.
Lastly, it is core team's responsibility to ensure that the PR meets all the SRU criteria.

> Point Release must follow the [Stable Release Policy](./STABLE_RELEASES.md).

Author: Right, I will not use _point release_.

Submitting a pull request:
- before submitting a pull request, merge the latest master `git merge origin/master`,
- run `make lint test` to ensure that all checks and tests pass,
- GitHub will squash commits and rebase on top of the master.

Finalizing pull requests
- ask contributors for review,
- respond to all comments,
- resolve merge conflicts

Pull requests are merged automatically using `automerge` action.

### Major Release Procedure

A _major release_ is an increment of the first number (eg: `v1.2` → `v2.0.0`) or the point number (eg: `v1.1 → v1.2.0`).

TL;DR before making a new _major_ release we do beta and release candidate releases. For example, for release 1.0.0:
```
v1.0.0-beta1 → v1.0.0-beta2 → ... → v1.0.0-rc1 → v1.0.0-rc2 → ... → v1.0.0
```

- Release a first beta version on `master` branch and freeze `master` for receiving any new features. Once beta is released, we focus on releasing the release candidate:
  - finish audits and reviews
  - kick off a large round of simulation testing (e.g. 400 seeds for 2k blocks)
  - perform functional tests
  - add more tests
  - release new beta version as the bugs are discovered and fixed.

> - Once the team feels that the `master` works fine we create a `release/vY` branch (going forward known a release branch), where `Y` is the version number, with the patch part substituted to `x` (eg: 0.42.x, 1.0.x). Ensure the release branch is protected against pushing from anyone except the release manager/coordinator.

Reviewer 2: ```suggestion
- After the team feels that the `master` works fine we create a `release/vY` branch (going forward known a release branch), where `Y` is the version number, with the patch part substituted to `x` (eg: 0.42.x, 1.0.x). Ensure the release branch is protected so that pushes against the release branch are permitted only by the release manager or release coordinator.
```

>     - **no PRs targeting this branch should be merged unless exceptional circumstances arise**

Reviewer 2: ```suggestion
    - **PRs targeting this branch can be merged _only_ when exceptional circumstances arise**
```

    - update the GitHub mergify integration by adding instructions for automatically backporting commits from `master` to the `release/vY` using the `backport/Y` label.
- In the release branch, prepare a new version section in the `CHANGELOG.md`
    - All links must be link-ified: `$ python ./scripts/linkify_changelog.py CHANGELOG.md`
    - Copy the entries into a `RELEASE_CHANGELOG.md`, this is needed so the bot knows which entries to add to the release page on GitHub.
- Create a new annotated git tag for a release candidate  (eg: `git tag -a v1.1.0-rc1`) in the release branch.
  - from this point we unfreeze master.

>   - the SDK teams should collaborate and do their best to run test nets in order to validate the release.
>   - once bugs are found, create a PR for `master` and backport to the release branch.
>   - create new release candidate tags once bugs are fixed.
> - Once the team feels the release branch is stable and everything works, create a full release:
>   - update `CHANGELOG.md`.
>   - create a new annotated git tag (eg `git -a v1.1.0`) in the release branch.

Reviewer 2: ```suggestion
  - the SDK teams collaborate and do their best to run testnets in order to validate the release.
  - when bugs are found, create a PR for `master`, and backport fixes to the release branch.
  - create new release candidate tags after bugs are fixed.
- After the team feels the release branch is stable and everything works, create a full release:
  - update `CHANGELOG.md`.
  - create a new annotated git tag (eg `git -a v1.1.0`) in the release branch.
```

Reviewer 2: another word that we can watch for is "once" let's use "after" in almost all cases, see https://developers.google.com/style/word-list?hl=en#letter-o 

> **once**
> If you mean after, then use after instead of once.

  - Create a GitHub release.

### Minor Release Procedure

A _minor release_ is an increment of the patch number (eg: `v1.2.0` → `v1.2.1`).

> Following _semver_ philosophy, we point release after `v1.0` must not break API but they may break consensus.

Reviewer 2: ```suggestion
Following _semver_ philosophy, point releases after `v1.0`:

 - must not break API
 - can break consensus
```

Reviewer 2: I think we're saying that it's permissible for consensus break? in this is true, let's use "can"

Reviewer 2: may
In general, reserve for official policy or legal considerations. To convey possibility, use can or might instead. **To convey permission, use can instead.**

Author: > I think we're saying that it's permissible for consensus break?

yes

Patch release must not break API nor consensus.
Before `v1.0`, point release can break both point API and consensus.

> Updates to the release branch should come from `master` by backporting PRs (usually done by automatic cherry pick followed by a PRs to the release branch). The backports must be marked using `backport/Y` label in PR for master.

Reviewer 2: ```suggestion
Updates to the release branch come only from `master` by backporting PRs (usually done by automatic cherry-pick followed by PRs to the release branch). The backports must be marked using `backport/Y` label in PR for master.
```

Author: In fact the original text is correct. "Should" has to be used. Updates which are not related to the upstream, must be created as PR to the release branch.

Reviewer 2: okay, not critical for me... although _should_ is ambiguous... I think we're saying that 

> Updates to the release branch can be made only from master by... 

(isn't this fun? I can talk word choice all day long, love this stuff!)

> It is the PR's author's responsibility to fix merge conflicts, update changelog entries, and

Reviewer 2: ```suggestion
It is the PR author's responsibility to fix merge conflicts, update changelog entries, and
```

> ensure CI passes. If a PR originates from an external contributor, it may be a core team member's

Reviewer 2: ```suggestion
ensure CI passes. If a PR originates from an external contributor, a core team member assumes 
```

responsibility to perform this process instead of the original author.
Lastly, it is core team's responsibility to ensure that the PR meets all the SRU criteria.

Point Release must follow the [Stable Release Policy](./STABLE_RELEASES.md).

> Once the release branch has all commits required for the next minor release:

Reviewer 2: ```suggestion
After the release branch has all commits required for the next minor release:
```

> means guaranteed by any quantifiable metric. It is a symbol of great trust of

Reviewer 2: ```suggestion
means guaranteed by any quantifiable metric. Serving as a code owner is a symbol of great trust from 
```

Reviewer 2: let's state the subject here, since the naked "it" allows ambiguity (especially for this awesome message)
I'm on a mission to remove all unclear references :), although "it" isn't clearly included in a developer documentation style guide, the same principle of "the naked it" applies to:

> this, that
> Where possible, put a noun after this or that for clarity. If doing so results in clunky prose, then don't do it; but even then, try thinking about what the noun would be. If you aren't sure what noun this or that (or in this case "It") refers to, then consider rephrasing.

* Code reviewers should have more senior engineering capability
* 1/2 approval is required from the **primary repo maintainers** in `CODEOWNERS`

> **Note**: For any major release series denoted as a "Stable Release" (e.g. v0.39 "Launchpad"), a separate release

Reviewer 3: ```suggestion
**Note**: For any major release series denoted as a "Stable Release" (e.g. v0.42 "Stargate"), a separate release
```
